src/File/RenameUpload.php

```diff
@@ -171,32 +173,49 @@ public function filter($value)
             }
 
             $isFileUpload = true;
-            $uploadData = $value;
             $sourceFile = $value['tmp_name'];
+            $clientFilename = $value['name'];
+        } elseif ($value instanceof UploadedFileInterface) {
+            $isFileUpload = true;
+            $sourceFile = $value->getStream()->getMetadata('uri');
+            $clientFilename = $value->getClientFilename();
         } else {
-            $uploadData = [
-                'tmp_name' => $value,
-                'name'     => $value,
-            ];
+            $clientFilename = $value;
             $sourceFile = $value;
         }
 
         if (isset($this->alreadyFiltered[$sourceFile])) {
             return $this->alreadyFiltered[$sourceFile];
         }
 
-        $targetFile = $this->getFinalTarget($uploadData);
-        if (! file_exists($sourceFile) || $sourceFile == $targetFile) {
+        $targetFile = $this->getFinalTarget($sourceFile, $clientFilename);
+        if ($sourceFile === $targetFile || ! file_exists($sourceFile)) {
             return $value;
         }
 
         $this->checkFileExists($targetFile);
-        $this->moveUploadedFile($sourceFile, $targetFile);
+        if ($value instanceof UploadedFileInterface) {
+            $value->moveTo($targetFile);
+        } else {
+            $this->moveUploadedFile($sourceFile, $targetFile);
+        }
 
         $return = $targetFile;
         if ($isFileUpload) {
-            $return = $uploadData;
-            $return['tmp_name'] = $targetFile;
+            if ($value instanceof UploadedFileInterface) {
+                $return = new UploadedFile(
```

Comment:
We should rely on PSR-17 here, instead of requiring Diactoros; that way any PSR-7 implementation can be used.

Doing so will require:

- Accepting a PSR-17 `UploadedFileFactoryInterface` as a filter option.
- Updating the tests to inject a mock PSR-17 `UploadedFileFactoryInterface` in the filter, to test that it is used to create a new instance.

I'll start working on those changes immediately.